### PR TITLE
Support spec.deployments to override configuration of system deployments

### PR DESCRIPTION
**Pull request by nak3**
Current spec has some problems:

- Annotations and Labels cannot be added safely into system pods.
- HA Logic is obsolete.
- HA setting applies to the all system pods. e.g. HA=2 means all deployments (webhook,controller,etc) scale to 2.
- HA setting does not work for Ingress manifests like Kourier and Contour.
- Resource setting does not work accurately if there are two pods have a same name container. e.g. webhook and istio-webhook have `webhook` container.
  ...etc.

To fix them, this patch introduces `deployments` field to override HA and Label/Annitations field for deployments.

Fixes https://github.com/knative/operator/issues/442 
~https://github.com/knative/operator/issues/382~ ~https://github.com/knative/operator/issues/337~

e.g. Override webhook and 3scale-kourier-controller

```
cat <<-EOF | kubectl apply -f -
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: ks
  namespace: knative-serving
spec:
  high-availability:
    replicas: 1
  deployments:
  - name: webhook # Override webhook deployments.
    replicas: 2
    labels:
      foo: bar
  - name: 3scale-kourier-control
    replicas: 2
    labels:
      foo: bar
  ingress:
    istio:
      enabled: false
    kourier:
      enabled: true
  config:
    network:
      ingress.class: kourier.ingress.networking.knative.dev
EOF
```

**Release Note**

```release-note
Support `spec.deployments` to override configuration of system deployments such as label,annotation and replicas.
```

### CODE REVIEW

**Comment by matzew** [comment]
Is there an issue captured for documenting these changes on `knative/docs` repo?

**Comment by houshengbo** [comment]
@nak3 I would suggest NOT deprecate the spec.HA.
We can leave it as an option to configure all the deployments with the same number of replicas.
spec.deployment to overwrite if available.

**Comment by nak3** [comment]
@matzew I haven't filed the change on `knative/docs` repo yet because I was (am still) not sure if this change will be accepted or not :worried:

@houshengbo Okay, I understand about spec.HA.
So, if we have the following configuration, `webhook` has `3` replicas, but other deployments have `2` replicas. Is that alright?

```
spec:
  high-availability:
    replicas: 2
  deployments:
  - name: webhook
    replicas: 3
```

For `spec.resources`, okay I will drop the change now.
But https://github.com/knative/operator/issues/382#issuecomment-741002226 needs to be solved somehow. istio-webhook and webhook already hit the issue.

**Comment by houshengbo** [comment]
@nak3 Right, that's how we can plan the usage of the ha and the deployment replicas together.

File: pkg/reconciler/common/deployments_override.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	if deployment.Spec.Template.GetLabels() == nil {
+		deployment.Spec.Template.Labels = map[string]string{}
+	}
+	for key, val := range override.Labels {
+		deployment.Labels[key] = val
+		deployment.Spec.Template.Labels[key] = val
+	}
+}
+
+func replaceReplicas(u *unstructured.Unstructured, override *v1alpha1.DeploymentOverride) error {
```
Comment (houshengbo) [inline pkg/reconciler/common/deployments_override.go:86]: Could you help to refactor this function into changing the deployment like replaceLabels?
Let's say we get rid of setting fields in the unstructured, but changing the fields in the deployment object.

Comment (nak3) [reply pkg/reconciler/common/deployments_override.go]: Sure, updated. Thank you.

**Pull request merged**